dig/dig.go

```diff
@@ -0,0 +1,288 @@
+func (g *graph) Resolve(obj interface{}) error {
+	objType := reflect.TypeOf(obj)
+	if objType.Kind() != reflect.Ptr {
+		return fmt.Errorf("can not resolve non-pointer object of type %v", objType)
+	}
+
+	objElemType := reflect.TypeOf(obj).Elem()
+	objVal := reflect.ValueOf(obj)
+
+	// check if the type has been nodes
+	n, ok := g.nodes[objElemType]
+	if !ok {
+		return fmt.Errorf("type %v is not registered", objType)
+	}
+
+	// check that all the dependencies have nodes present in the graph
+	// doesn't mean everything will go smoothly during resolve, but it
+	// drastically increases the chances that we're not missing something
+	for _, node := range g.nodes {
+		for _, dep := range node.dependencies() {
+			// check that the dependency has been nodes
+			_, ok := g.nodes[dep]
+			if !ok {
+				return fmt.Errorf("%v dependency of type %v is not registered", objElemType, dep)
+			}
+		}
+	}
+
+	v, err := n.value(g)
+	if err != nil {
+		return errors.Wrapf(err, "unable to resolve %v", objType)
+	}
+
+	// set the pointer value of the provided object to the instance pointer
+	objVal.Elem().Set(v)
+
+	return nil
+}
+
+type object interface {
+	// Return value of the object
+	value(g *graph) (reflect.Value, error)
+
+	// Other things that need to be present before this object can be created
+	dependencies() []interface{}
+}
+
+type graph struct {
+	fmt.Stringer
+
+	nodes map[interface{}]object
+}
+
+func newGraph() *graph {
+	return &graph{
+		nodes: make(map[interface{}]object),
+	}
+}
+
+func (g graph) String() string {
+	var b bytes.Buffer
+	for key, reg := range g.nodes {
+		b.WriteString(fmt.Sprintf("%v -> %v\n", key, reg))
+	}
+	return fmt.Sprintf("{nodes:\n%v}", b.String())
+}
+
+type node struct {
+	fmt.Stringer
+
+	obj     interface{}
+	objType reflect.Type
+	cached  bool
+}
+
+// Return the earlier provided instance
+func (n node) value(g *graph) (reflect.Value, error) {
+	return reflect.ValueOf(n.obj), nil
+}
+
+func (n node) dependencies() []interface{} {
+	return nil
+}
+
+func (n node) String() string {
+	return fmt.Sprintf(
+		"(object) obj: %v, deps: nil, cached: %v",
+		n.objType,
+		n.cached,
+	)
+}
+
+type funcNode struct {
+	node
+	fmt.Stringer
+
+	constructor interface{}
+	deps        []interface{}
+}
+
+// Call the function and return the result
+func (n funcNode) value(g *graph) (reflect.Value, error) {
+	cv := reflect.ValueOf(n.constructor)
+	ct := reflect.TypeOf(n.constructor)
+
+	args := make([]reflect.Value, ct.NumIn(), ct.NumIn())
+	for idx := range args {
+		arg := ct.In(idx)
+		node, ok := g.nodes[arg]
+		if ok {
+			v, err := node.value(g)
+			if err != nil {
+				return reflect.Zero(n.objType), errors.Wrap(err, "dependency resolution failed")
+			}
+			args[idx] = v
+		}
+	}
+
+	l := list.New()
+	for _, node := range g.nodes {
+		for _, dep := range node.dependencies() {
+			node, ok := g.nodes[dep]
+			if ok {
+				l.PushFront(node)
+			}
+		}
+	}
+
+	return cv.Call(args)[0], nil
```

Comment:
If we force a `ptr,error` tuple for constructors, yes error would be `[1]` in the returns